### ISSUE TITLE
Should rest-client close payload streams?

### ISSUE THREAD

**Issue by ab** [rest-client/rest-client#441]
It's not clear that this should be rest-client's responsibility. It makes it very difficult to reuse the same file handle for e.g. a redirect.

This behavior seems to date to extremely early on, and carries no explanation. 8b000e1a602d005e20a79e6375493de1cb4683e7

**Comment by lagartoflojo** [comment]
Hi,
I'm having a problem because of the closed streams.
Basically, when RestClient raises an exception, I want to log the request and response data (in my particular case, using the Airbrake gem, but that's not really relevant) in order to figure out why the call to an external API failed.
So when a `RestClient::Exception` is raised, I rescue from it and then try to extract as much data from it as possible:

```ruby
def exception_payload(exception)
  # Timeout errors don't have a response object
  if exception.response && exception.response.request
    request = exception.response.request
    payload.merge!(
      url: request.url,
      method: request.method,
      request_body: request.payload.to_s,
      response_code: exception.http_code,
      response_body: exception.http_body,
      response_headers: exception.http_headers
    )
  end || {}
end
```

I can basically retrieve all the information I need from the exception's `response` and `request` objects, but reading the request's `payload` is not possible, because it has already been closed.
Of course, I have access to the payload from somewhere else (since I built it), but then I would always have to pass it to the `exception_payload` method, which feels unnecessary.

Any ideas? Thanks!

**Comment by ab** [comment]
Yeah, I'm pretty confused about what the original intent was behind explicitly closing payloads. But I hesitate to change the behavior without fully understanding the implications. Thoughts welcome!

**Comment by kwstannard** [comment]
I just ran into this problem. The retryable gem blew up with an IO error because RestClient closed the file.

RestClient should 100% not be altering the state of any objects passed to RestClient.
